File: modules/spring-security/testdata/java/references/UserDetailsConfiguration2.java
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.test.context.support.WithUserDetails;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Collections;


@Service
class UserDetailsServiceImpl implements UserDetailsService{
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        return null;
    }
}

@Configuration
public class ConfigurationResolve {
    @Bean
    UserDetailsServiceImpl admin(){
        return null;
    }

    @Bean("admin")
    UserDetailsService adminUserDetailsService() {
        return username -> {
            UserDetails userDetails = new UserDetails() {
                @Override
                public Collection<? extends GrantedAuthority> getAuthorities() {
                    return Collections.emptyList();
                }

                @Override
                public String getPassword() {
                    return "pass";
                }

                @Override
                public String getUsername() {
                    return "name";
                }

                @Override
                public boolean isAccountNonExpired() {
                    return false;
                }

                @Override
                public boolean isAccountNonLocked() {
                    return false;
                }

                @Override
                public boolean isCredentialsNonExpired() {
                    return false;
                }

                @Override
                public boolean isEnabled() {
                    return false;
                }
            };
            return userDetails;
        };
    }
}

File: modules/spring-security/testdata/kotlin/references/UserDetailsConfiguration2.kt
package src

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.test.context.support.WithUserDetails
import org.springframework.stereotype.Service


@Service
internal class KotlinUserDetailsServiceImpl : UserDetailsService {
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails? {
        return null
    }
}

@Configuration
internal open class KotlinUserDetailsConfiguration {
    @Bean
    open fun kotlinAdmin(): KotlinUserDetailsServiceImpl? {
        return null
    }

    @Bean("kotlinAdmin")
    open fun kotlinAdminUserDetailsService(): UserDetailsService {
        return UserDetailsService { _: String? ->
            val userDetails = object : UserDetails {
                override fun getAuthorities(): Collection<GrantedAuthority> {
                    return emptyList()
                }

                override fun getPassword(): String {
                    return "pass"
                }

                override fun getUsername(): String {
                    return "name"
                }

                override fun isAccountNonExpired(): Boolean {
                    return false
                }

                override fun isAccountNonLocked(): Boolean {
                    return false
                }

                override fun isCredentialsNonExpired(): Boolean {
                    return false
                }

                override fun isEnabled(): Boolean {
                    return false
                }
            }
            userDetails
        }
    }
}
